lib/PrintAsObjC/PrintAsObjC.cpp

```diff
@@ -867,13 +880,37 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
     //   NSDictionary<NSObject *, id> --> NSDictionary
     //   NSSet<id> --> NSSet
     if (!typeArgs.empty() &&
-        (!hasGenericObjCType(objcClass) ||
-         (swiftNominal == ctx.getArrayDecl() && typeArgs[0]->isAnyObject()) ||
-         (swiftNominal == ctx.getDictionaryDecl() &&
-          isNSObject(ctx, typeArgs[0]) && typeArgs[1]->isAnyObject()) ||
-         (swiftNominal == ctx.getSetDecl() && isNSObject(ctx, typeArgs[0])))) {
+        (!hasGenericObjCType(objcClass)
+         || (swiftNominal == ctx.getArrayDecl() &&
+             isAnyObjectOrAny(typeArgs[0]))
+         || (swiftNominal == ctx.getDictionaryDecl() &&
+             isNSObjectOrAnyHashable(ctx, typeArgs[0]) &&
+             isAnyObjectOrAny(typeArgs[1]))
+         || (swiftNominal == ctx.getSetDecl() &&
+             isNSObjectOrAnyHashable(ctx, typeArgs[0])))) {
       typeArgs = {};
     }
+    
+    // Use the proper upper id<NSCopying> bound for Dictionaries with
+    // upper-bounded keys.
+    else if (swiftNominal == ctx.getDictionaryDecl() &&
+             isNSObjectOrAnyHashable(ctx, typeArgs[0])) {
+      if (Module *M = ctx.getLoadedModule(ctx.Id_Foundation)) {
+        UnqualifiedLookup lookup(ctx.getIdentifier("NSCopying"), M, nullptr);
```

Comment:
`lookupQualified` is probably more appropriate here. Also, the `if (!NSCopyingType)` should guard this whole section.